### ISSUE TITLE
cqlsh can't connect to cassandra on pure IPv6 network when rpc_address is defaulted

### ISSUE THREAD

**Issue by dtheobald** [Metaswitch/clearwater-cassandra#66]
On a pure IPv6 network (i.e. split network with IPv4 management and IPv6 signaling), ```cqlsh``` can't connect to ```cassandra``` if its ```rpc_address``` is set to ```localhost``` (the default). When I specifically configured cassandra to used the IPv6 ```localhost``` address (```0:0:0:0:0:0:0:1```) for its ```rpc_address```, ```cqlsh``` could then connect if it also used ```0:0:0:0:0:0:0:1```.

I've temporarily fixed this with this change:
```
--- a/clearwater-cassandra/usr/share/clearwater/clearwater-cluster-manager/plugins/cassandra_plugin.py
+++ b/clearwater-cassandra/usr/share/clearwater/clearwater-cluster-manager/plugins/cassandra_plugin.py
@@ -42,6 +42,7 @@ import time
 import yaml
 import os
 import subprocess
+import socket
 
 _log = logging.getLogger("cassandra_plugin")
 
@@ -74,6 +75,16 @@ def join_cassandra_cluster(cluster_view,
         # Fill in the correct listen_address and seeds values in the yaml
         # document.
         doc["listen_address"] = ip
+
+        # Fix up cassandra's rpc_address for IPv4/6. On a pure IPv6
+        # network cassandra/cqlsh won't work properly using the
+        # default value of "localhost"
+        try:
+            socket.inet_pton(socket.AF_INET6, ip)
+            doc["rpc_address"] = "0:0:0:0:0:0:0:1"
+        except socket.error:
+            doc["rpc_address"] = "127.0.0.1"
+
```

**Comment by AlexHockey** [comment]
I've spun up a deployment with a pure IPv6 signaling network. These are the notes I've taken so far.

In order to get cassandra to start at all, you need to apply the workaround in https://github.com/Metaswitch/clearwater-cassandra/issues/67. This gets the inter-cassandra port (port 7000) bound the signaling local IP address (which is an IPv6 address) and leaves the RPC port (9160) and JMX port (7199) on 127.0.0.1. 

The JMX port seems happy being on 127.0.0.1 as nodetool status works fine. 

However in a pure IPv6 network cqlsh needs to connect to an IPv6 address (see https://github.com/Metaswitch/homestead/issues/13). I tried setting `rpc_address` to ::1. This got the RPC port onto ::1 (I'm not sure why this didn't work for @dtheobald). However cqlsh still couldn't connect however because localhost did not resolve to ::1. I've found two ways to make it connect:

* Supply the host on the command line: `sudo ip netns exec signaling cqlsh ::1`. This is a bit annoying as it means we would have to update all our invocations of cqlsh.
* Adding `::1 locahost` to `/etc/hosts` but I'm not sure this is the right fix as Ubuntu doesn't set up this mapping by default.

For the time being I've got my system set up with the extra entry in `/etc/hosts` so I can make progress and flush out any remaining issues. 

This flushed out an issue that `poll_cassandra.sh` uses "127.0.0.1" as the host to connect to. 

I now can connect to cassandra using cqlsh and cassandra-cli, but homestead-prov isn't working yet. Need to look into that next.

**Comment by dtheobald** [comment]
I don't know why ```::1``` didn't work for me, but I was only hacking to get myself unblocked, so once I got something that basically worked, I put my machete away.

**Comment by dtheobald** [comment]
BTW - I set ```cassandra_hostname``` to the IPv6 localhost, so as to not have to change all invocations of ```cqlsh```

**Comment by AlexHockey** [comment]
I've commit some fixes for this in https://github.com/Metaswitch/clearwater-etcd/pull/245 and #70. 

Note that even with these fixes (and mapping localhost -> ::1) homestead still doesn't install cleanly on IPv6, nor does it work when I try and fix stuff up manually, but I'm going to track those as a separate issue.
